shell/platform/android/io/flutter/embedding/android/FlutterFragment.java

```diff
@@ -258,55 +261,63 @@ private void initializeFlutter(@NonNull Context context) {
-  protected FlutterEngine onCreateFlutterEngine(@NonNull Context context) {
-    Log.d(TAG, "onCreateFlutterEngine()");
-    return new FlutterEngine(context);
-  }
+  protected void setupFlutterEngine() {
+    // First, defer to the FragmentActivity that owns us to see if it wants to provide a
+    // FlutterEngine.
+    FragmentActivity attachedActivity = getActivity();
+    if (attachedActivity instanceof FlutterEngineProvider) {
+      // Defer to the Activity that owns us to provide a FlutterEngine.
+      Log.d(TAG, "Deferring to attached Activity to provide a FlutterEngine.");
+      FlutterEngineProvider flutterEngineProvider = (FlutterEngineProvider) attachedActivity;
+      flutterEngine = flutterEngineProvider.getFlutterEngine(getContext());
+      if (flutterEngine != null) {
+        isFlutterEngineFromActivity = true;
+      }
+    }
 
-  /**
-   * Hook for subclasses to customize the {@link FlutterEngine} owned by this {@link FlutterFragment}
-   * after the {@link FlutterEngine} has been instantiated.
-   *
-   * Consider using this method to connect desired Flutter plugins to this {@code Fragment}'s
-   * {@link FlutterEngine}.
-   */
-  protected void onFlutterEngineCreated(@NonNull FlutterEngine flutterEngine) {
-    // no-op
+    // If flutterEngine is null then either our Activity is not a FlutterEngineProvider,
+    // or our Activity decided that it didn't want to provide a FlutterEngine. Either way,
+    // we will now create a FlutterEngine for this FlutterFragment.
+    if (flutterEngine == null) {
+      // Create a FlutterEngine to back our FlutterView.
+      Log.d(TAG, "Our attached Activity did not want to provide a FlutterEngine. Creating a "
+          + "new FlutterEngine for this FlutterFragment.");
+      flutterEngine = new FlutterEngine(getContext());
+      isFlutterEngineFromActivity = false;
+    }
   }
 
   @Nullable
   @Override
   public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
     flutterView = new FlutterView(getContext(), getRenderMode());
-    flutterView.attachToFlutterEngine(flutterEngine);
 
-    // TODO(mattcarroll): the following call should exist here, but the plugin system needs to be revamped.
-    //                    The existing attach() method does not know how to handle this kind of FlutterView.
-    //flutterEngine.getPluginRegistry().attach(this, getActivity());
-
-    doInitialFlutterViewRun();
+    // We post() the code that attaches the FlutterEngine to our FlutterView because there is
+    // some kind of blocking logic on the native side when the surface is connected. That lag
+    // causes launching Activitys to wait a second or two before launching. By post()'ing this
+    // behavior we are able to move this blocking logic to after the Activity's launch.
+    // TODO(mattcarroll): figure out how to avoid blocking the MAIN thread when connecting a surface
+    new Handler().post(new Runnable() {
```

Comment:
I don't think so.  I've never inspected the result of `post()` before, and the only explicit cause of this eventuality from the docs says "usually because the looper processing the message queue is exiting." But we're attaching to the MAIN thread, which won't exit until the application dies. So I think we're OK.